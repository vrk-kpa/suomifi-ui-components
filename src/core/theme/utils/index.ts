export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
