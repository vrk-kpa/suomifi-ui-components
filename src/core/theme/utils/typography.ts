import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
