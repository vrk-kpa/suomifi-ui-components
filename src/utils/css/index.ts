export { alphaHex, hslaToHex } from './colors';
export { allStates, absolute } from './pseudo';
export {
  clearfix,
  cssObjectToCss,
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from './utils';
