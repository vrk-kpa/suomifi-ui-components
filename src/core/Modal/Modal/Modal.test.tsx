import React, { ReactNode, useRef, useState } from 'react';
import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { axeTest } from '../../../utils/test';

import { Modal, ModalProps } from './Modal';
import { ModalContent, ModalFooter, ModalTitle } from '../';
import { Button } from '../../../';
import { createRoot, Root } from 'react-dom/client';

let appRoot: HTMLDivElement | null = null;
let root: Root | null = null;

beforeEach(() => {
  appRoot = document.createElement('div');
  appRoot.setAttribute('id', 'root');
  document.body.appendChild(appRoot);
  root = createRoot(appRoot);
});

afterEach(() => {
  if (root) {
    act(() => {
      root?.unmount();
    });
    root = null;
  }
  if (appRoot) {
    appRoot.remove();
    appRoot = null;
  }
});

describe('Basic modal', () => {
  const text = 'Modal Content';
  const BasicModal = (props?: Partial<ModalProps>) => (
    <Modal
      appElementId="root"
      visible={true}
      {...props}
      ariaLabelledBy="test12"
    >
      <ModalContent>
        <ModalTitle>Test modal</ModalTitle>
        <p>{text}</p>
      </ModalContent>
      <ModalFooter>
        <Button>OK</Button>
      </ModalFooter>
    </Modal>
  );

  it('should be hidden when visible is set to false', () => {
    const { baseElement } = render(BasicModal({ visible: false }));
    expect(baseElement.querySelector('[class~="fi-modal"]')).toBe(null);
  });

  it('should have given className', () => {
    const { getByRole } = render(BasicModal({ className: 'test-class-name' }));
    expect(getByRole('dialog')).toHaveClass('test-class-name');
  });

  it('should match snapshot', () => {
    const { baseElement } = render(BasicModal());
    expect(baseElement).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', () => {
    const { baseElement } = render(BasicModal());
    axeTest(baseElement as any);
  });

  it('should have given aria-labelledby', () => {
    const { getByRole } = render(BasicModal());
    expect(getByRole('dialog')).toHaveAttribute('aria-labelledby', 'test12');
  });
});

describe('Modal sibling DOM nodes', () => {
  const ModalWithSiblings = () => {
    const [open, setOpen] = useState(true);
    return (
      <Modal appElementId="root" visible={open}>
        <ModalContent>
          <ModalTitle>Test modal</ModalTitle>
          <p>Some test text</p>
        </ModalContent>
        <ModalFooter>
          <Button onClick={() => setOpen(false)}>OK</Button>
        </ModalFooter>
      </Modal>
    );
  };

  it('containing the modal should not be aria-hidden', () => {
    const { baseElement } = render(<ModalWithSiblings />);
    const modalRoot = baseElement.querySelector('[class~="fi-modal_base"]');
    expect(modalRoot).not.toHaveAttribute('aria-hidden');
  });

  it('not containing the modal should be aria-hidden', () => {
    const { baseElement } = render(<ModalWithSiblings />);
    const appRootNode = baseElement.querySelector('#root');
    expect(appRootNode).toHaveAttribute('aria-hidden', 'true');
  });

  it('should preserve aria-hidden and role state after closing modal', () => {
    const { getByRole, baseElement } = render(<ModalWithSiblings />);
    fireEvent.click(getByRole('button'));
    const appRootNode = baseElement.querySelector('#root');
    expect(appRootNode).not.toHaveAttribute('aria-hidden');
  });
});

describe('Modal variant', () => {
  const ModalWithProps = (props?: Partial<ModalProps>) => (
    <Modal appElementId="root" visible={true} {...props}>
      <ModalContent>
        <ModalTitle>Test modal</ModalTitle>
        <p>Modal Content</p>
      </ModalContent>
      <ModalFooter>
        <Button>OK</Button>
        <Button variant="secondary">Cancel</Button>
      </ModalFooter>
    </Modal>
  );

  it('smallScreen should have correct classname', () => {
    const { getByRole } = render(ModalWithProps({ variant: 'smallScreen' }));
    expect(getByRole('dialog')).toHaveClass('fi-modal--small-screen');
  });
});

describe('Modal style', () => {
  it('should have given inline styles', () => {
    const { getByRole } = render(
      <Modal appElementId="root" visible={true} style={{ width: '1000px' }}>
        <ModalContent>
          <ModalTitle>Test modal</ModalTitle>
          <p>Modal Content</p>
        </ModalContent>
        <ModalFooter>
          <Button>OK</Button>
          <Button variant="secondary">Cancel</Button>
        </ModalFooter>
      </Modal>,
    );
    expect(getByRole('dialog')).toHaveStyle('width: 1000px');
  });
});

describe('Modal focus', () => {
  const ModalWithProps = (children: ReactNode, props?: Partial<ModalProps>) => (
    <Modal appElementId="root" visible={true} {...props}>
      <ModalContent>
        <ModalTitle>Test modal</ModalTitle>
        {children}
      </ModalContent>
      <ModalFooter>
        <Button>OK</Button>
        <Button variant="secondary">Cancel</Button>
      </ModalFooter>
    </Modal>
  );

  it('should be on heading by default', async () => {
    const { getAllByRole } = render(ModalWithProps(<p>Modal Content</p>));
    await waitFor(() => expect(getAllByRole('heading')[0]).toHaveFocus());
  });

  it('should be on element provided in props', async () => {
    const ModalWithFocusOnOpenRef = (props?: Partial<ModalProps>) => {
      const buttonRef = React.createRef<HTMLButtonElement>();
      return (
        <Modal
          appElementId="root"
          visible={true}
          focusOnOpenRef={buttonRef}
          {...props}
        >
          <ModalContent>
            <ModalTitle>Test modal</ModalTitle>
            <button>Test button 1</button>
            <button ref={buttonRef}>Test button 2</button>
          </ModalContent>
          <ModalFooter>
            <Button>OK</Button>
            <Button variant="secondary">Cancel</Button>
          </ModalFooter>
        </Modal>
      );
    };
    const { getByText } = render(<ModalWithFocusOnOpenRef />);
    await waitFor(() => expect(getByText('Test button 2')).toHaveFocus());
  });
});

describe('Closing Modal', () => {
  const RefTest = (props?: Partial<ModalProps>) => {
    const [open, setOpen] = useState(false);
    const ref = useRef(null);
    return (
      <>
        <button onClick={() => setOpen(!open)}>Toggle modal</button>
        <input ref={ref} type="text" />
        <Modal
          appElementId="root"
          visible={open}
          focusOnCloseRef={ref}
          {...props}
        >
          <ModalContent>
            <ModalTitle>Test modal</ModalTitle>
            Test Content
            <button onClick={() => setOpen(false)}>Close modal</button>
          </ModalContent>
          <ModalFooter>
            <Button>OK</Button>
            <Button variant="secondary">Cancel</Button>
          </ModalFooter>
        </Modal>
      </>
    );
  };

  it('should be possible with ESC key', () => {
    const mockEsc = jest.fn();
    const { getByText } = render(
      <RefTest
        focusOnCloseRef={undefined}
        onEscKeyDown={() => {
          mockEsc();
        }}
      />,
    );
    const openButton = getByText('Toggle modal');
    fireEvent.click(openButton);
    const content = getByText('Test Content');
    expect(content).toBeTruthy();

    fireEvent.keyDown(content, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
    });

    expect(mockEsc).toHaveBeenCalledTimes(1);
  });

  it('should return focus to original element', async () => {
    const { getByText } = render(<RefTest focusOnCloseRef={undefined} />);
    const openButton = getByText('Toggle modal');
    openButton.focus();
    fireEvent.click(openButton);
    const closeButton = getByText('Close modal');
    expect(closeButton).toBeTruthy();
    fireEvent.click(closeButton);
    await waitFor(() => expect(openButton).toHaveFocus());
  });

  it('should return focus to given element', async () => {
    const { getByText, getByRole } = render(<RefTest />);
    const openButton = getByText('Toggle modal');
    fireEvent.click(openButton);
    const closeButton = getByText('Close modal');
    expect(closeButton).toBeTruthy();
    fireEvent.click(closeButton);
    const input = getByRole('textbox');
    await waitFor(() => expect(input).toHaveFocus());
  });
});
