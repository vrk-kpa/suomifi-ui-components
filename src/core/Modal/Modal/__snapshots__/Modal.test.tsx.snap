// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic modal should match snapshot 1`] = `
.c0.fi-modal_base {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c0.fi-modal_base .fi-modal_overlay {
  background-color: rgba(33,33,33,0.5);
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.c0 .fi-modal {
  border-radius: 4px;
  background-color: hsl(0, 0%, 100%);
  border-top: 4px solid hsl(212, 63%, 45%);
  overflow: hidden;
  max-height: calc(100% - 50px);
  min-height: 230px;
  width: 800px;
  max-width: 100%;
  flex: 0 1 auto;
  display: flex;
  flex-direction: column;
  align-items: stretch;
}

.c0 .fi-modal:focus {
  outline: none;
}

.c0 .fi-modal:focus-visible {
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196, 77%, 44%);
  outline: 0;
}

.c0 .fi-modal--no-scroll {
  width: 540px;
  min-height: 190px;
}

.c0 .fi-modal--small-screen {
  border-radius: 0;
  width: 100%;
  height: 100%;
  max-height: 100%;
}

.c9 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c9:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c9::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c9::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c9::-webkit-inner-spin-button {
  height: auto;
}

.c9::-webkit-outer-spin-button {
  height: auto;
}

.c9 ::before,
.c9 ::after {
  box-sizing: border-box;
}

.c1 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c1 ::before,
.c1 ::after {
  box-sizing: border-box;
}

.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c5 ::before,
.c5 ::after {
  box-sizing: border-box;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c3 ::before,
.c3 ::after {
  box-sizing: border-box;
}

.c2 {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c2.fi-modal_content {
  flex: 1 1 auto;
  max-height: 100%;
  padding: 24px 30px 50px 30px;
  overflow-y: auto;
  scroll-padding-bottom: 75px;
  scroll-padding-top: 75px;
}

.c2.fi-modal_content--no-scroll {
  overflow: hidden;
  padding: 34px 30px 10px 30px;
}

.c2.fi-modal_content--small-screen {
  padding-top: 20px;
  padding-left: 15px;
  padding-bottom: 50px;
  padding-right: 15px;
}

.c6 {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(0, 0%, 13%);
}

.c6.fi-heading--h1hero {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 40px;
  line-height: 48px;
  font-weight: 600;
}

.c6.fi-heading--h1 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 40px;
  line-height: 48px;
  font-weight: 300;
}

.c6.fi-heading--h2 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 28px;
  line-height: 34px;
  font-weight: 600;
}

.c6.fi-heading--h3 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 22px;
  line-height: 28px;
  font-weight: 600;
}

.c6.fi-heading--h4 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 24px;
  font-weight: 600;
}

.c6.fi-heading--h5 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c6.fi-heading--h6 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c6.fi-heading--small-screen {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 32px;
  line-height: 38px;
  font-weight: 300;
}

.c6.fi-heading--small-screen.fi-heading--h1hero {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 32px;
  line-height: 38px;
  font-weight: 600;
}

.c6.fi-heading--small-screen.fi-heading--h1 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 32px;
  line-height: 38px;
  font-weight: 300;
}

.c6.fi-heading--small-screen.fi-heading--h2 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 26px;
  line-height: 32px;
  font-weight: 600;
}

.c6.fi-heading--small-screen.fi-heading--h3 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 20px;
  line-height: 26px;
  font-weight: 600;
}

.c6.fi-heading--small-screen.fi-heading--h4 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c6.fi-heading--small-screen.fi-heading--h5 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c6.fi-heading--small-screen.fi-heading--h6 {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c4 {
  max-width: 100%;
}

.c4.fi-modal_title {
  display: block;
  margin-bottom: 20px;
}

.c4.fi-modal_title .fi-modal_title_focus-wrapper {
  position: relative;
  display: inline-block;
  max-width: 100%;
}

.c4.fi-modal_title .fi-modal_title_focus-wrapper:focus {
  outline: 0;
}

.c4.fi-modal_title .fi-modal_title_focus-wrapper:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0, 0%, 100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196, 77%, 44%);
  z-index: 9999;
}

.c4.fi-modal_title--small-screen {
  margin-bottom: 20px;
}

.c7.fi-modal_footer {
  flex: 0 0 auto;
  position: relative;
}

.c7.fi-modal_footer .fi-modal_footer_content {
  padding-top: 0;
  padding-right: 15px;
  padding-bottom: 20px;
  padding-left: 30px;
}

.c7.fi-modal_footer .fi-modal_footer_content>* {
  margin-top: 20px;
  margin-right: 15px;
}

.c7.fi-modal_footer .fi-modal_footer_content-gradient-overlay {
  position: absolute;
  height: 51px;
  width: 100%;
  top: -51px;
  border-bottom: 1px solid hsl(202, 7%, 80%);
}

.c7.fi-modal_footer .fi-modal_footer_content-gradient-overlay .fi-modal_footer_content-gradient {
  height: 50px;
  margin: 0 20px 0 20px;
  background: linear-gradient(           0deg,           hsl(0, 0%, 100%),           rgba(255, 255, 255, 0)         );
}

.c7.fi-modal_footer--small-screen .fi-modal_footer_content {
  padding-right: 0;
  padding-bottom: 15px;
  padding-left: 15px;
}

.c7.fi-modal_footer--small-screen .fi-modal_footer_content>* {
  display: block;
  width: calc(100% - 15px);
  margin-top: 15px;
}

.c8 {
  position: absolute;
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

.c10 {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  padding: 9px 20px;
  min-height: 40px;
  color: hsl(0, 0%, 100%);
  background: linear-gradient(0deg, hsl(212, 63%, 37%) 0%, hsl(212, 63%, 45%) 100%);
  border-radius: 2px;
  text-align: center;
  text-shadow: 0 1px 1px hsla(214, 100%, 24%, 0.5);
  cursor: pointer;
  border: 1px solid transparent;
}

.c10:focus {
  position: relative;
  outline: 3px solid transparent;
}

.c10:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0, 0%, 100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196, 77%, 44%);
  z-index: 9999;
}

.c10:hover {
  background: linear-gradient(0deg, hsl(212, 63%, 45%) 0%, hsl(212, 63%, 49%) 100%);
  outline: 2px solid transparent;
}

.c10:active {
  background: hsl(212, 63%, 37%);
}

.c10.fi-button--disabled,
.c10[disabled],
.c10:disabled {
  text-shadow: 0 1px 1px rgba(33,33,33,0.5);
  background: linear-gradient(0deg, hsl(202, 7%, 67%) 0%, hsl(202, 7%, 80%) 100%);
  user-select: none;
  cursor: not-allowed;
}

.c10.fi-button--disabled::after {
  border: none;
  box-shadow: none;
}

.c10.fi-button--fullwidth {
  display: block;
  width: 100%;
}

.c10.fi-button--inverted {
  background: none;
  background-color: hsl(212, 63%, 45%);
  border: 1px solid hsl(0, 0%, 100%);
  text-shadow: none;
}

.c10.fi-button--inverted:hover {
  background: linear-gradient(-180deg, hsla(0, 0%, 100%, 0.1) 0%, hsla(0, 0%, 100%, 0) 100%);
}

.c10.fi-button--inverted:active {
  background: none;
  background-color: hsl(212, 63%, 45%);
}

.c10.fi-button--inverted.fi-button--disabled,
.c10.fi-button--inverted[disabled],
.c10.fi-button--inverted:disabled {
  opacity: 0.41;
  background: none;
  background-color: none;
}

.c10.fi-button--secondary {
  color: hsl(212, 63%, 45%);
  background: none;
  background-color: hsl(0, 0%, 100%);
  border: 1px solid hsl(212, 63%, 45%);
  text-shadow: none;
}

.c10.fi-button--secondary:hover {
  background: linear-gradient(-180deg, hsl(202, 7%, 93%) 0%, hsla(0, 0%, 100%, 0) 100%);
}

.c10.fi-button--secondary:active {
  background: none;
  background-color: hsl(202, 7%, 93%);
}

.c10.fi-button--secondary.fi-button--disabled,
.c10.fi-button--secondary[disabled],
.c10.fi-button--secondary:disabled {
  color: hsl(202, 7%, 67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212, 63%, 98%);
  border-color: hsl(202, 7%, 67%);
}

.c10.fi-button--secondary .fi-button_loading-icon .fi-icon-component-brand-fill {
  fill: hsl(202, 7%, 67%);
}

.c10.fi-button--secondary-noborder {
  color: hsl(212, 63%, 45%);
  background: none;
  background-color: hsl(0, 0%, 100%);
  border: 1px solid hsl(212, 63%, 45%);
  text-shadow: none;
  border: none;
  padding: 10px 20px;
  background-color: transparent;
}

.c10.fi-button--secondary-noborder:hover {
  background: linear-gradient(-180deg, hsl(202, 7%, 93%) 0%, hsla(0, 0%, 100%, 0) 100%);
}

.c10.fi-button--secondary-noborder:active {
  background: none;
  background-color: hsl(202, 7%, 93%);
}

.c10.fi-button--secondary-noborder.fi-button--disabled,
.c10.fi-button--secondary-noborder[disabled],
.c10.fi-button--secondary-noborder:disabled {
  color: hsl(202, 7%, 67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212, 63%, 98%);
  border-color: hsl(202, 7%, 67%);
}

.c10.fi-button--secondary-noborder .fi-button_loading-icon .fi-icon-component-brand-fill {
  fill: hsl(202, 7%, 67%);
}

.c10.fi-button--secondary-noborder.fi-button--icon-only {
  padding: 6px 12px;
}

.c10.fi-button--secondary-noborder .fi-button_loading-icon .fi-icon-component-brand-fill {
  fill: hsl(202, 7%, 67%);
}

.c10.fi-button--secondary-light {
  color: hsl(212, 63%, 45%);
  color: hsl(212, 63%, 45%);
  background: none;
  background-color: hsl(0, 0%, 100%);
  border: 1px solid hsl(212, 63%, 45%);
  text-shadow: none;
  background: linear-gradient(0deg, hsl(215, 100%, 95%), hsl(215, 100%, 97%));
  padding: 10px 20px;
  border: none;
}

.c10.fi-button--secondary-light:hover {
  background: linear-gradient(-180deg, hsl(202, 7%, 93%) 0%, hsla(0, 0%, 100%, 0) 100%);
}

.c10.fi-button--secondary-light:active {
  background: none;
  background-color: hsl(202, 7%, 93%);
}

.c10.fi-button--secondary-light.fi-button--disabled,
.c10.fi-button--secondary-light[disabled],
.c10.fi-button--secondary-light:disabled {
  color: hsl(202, 7%, 67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212, 63%, 98%);
  border-color: hsl(202, 7%, 67%);
}

.c10.fi-button--secondary-light .fi-button_loading-icon .fi-icon-component-brand-fill {
  fill: hsl(202, 7%, 67%);
}

.c10.fi-button--secondary-light.fi-button--icon-only {
  padding: 6px 12px;
}

.c10.fi-button--secondary-light:hover {
  background: linear-gradient(0deg, hsl(215, 100%, 97%), hsl(212, 63%, 98%));
}

.c10.fi-button--secondary-light:active {
  background: linear-gradient(0deg, hsl(202, 7%, 93%), hsl(202, 7%, 97%));
}

.c10.fi-button--secondary-light.fi-button--disabled,
.c10.fi-button--secondary-light[disabled],
.c10.fi-button--secondary-light:disabled {
  color: hsl(202, 7%, 67%);
  background: none;
  background-color: hsl(202, 7%, 97%);
}

.c10.fi-button--secondary-light .fi-button_loading-icon .fi-icon-component-brand-fill {
  fill: hsl(202, 7%, 67%);
}

.c10>.fi-button_icon>.fi-icon {
  width: 16px;
  height: 16px;
  margin-right: 8px;
  vertical-align: middle;
  transform: translateY(-0.1em);
}

.c10>.fi-button_icon--right>.fi-icon {
  margin-right: 0;
  margin-left: 8px;
}

.c10.fi-button--icon-only {
  padding: 6px 11px;
}

.c10.fi-button--icon-only>.fi-button_icon>.fi-icon {
  margin-right: 0;
}

.c10.fi-button--icon-only>.fi-button_icon--right>.fi-icon {
  margin-left: 0;
}

.c10.fi-button--disabled>.fi-button_icon {
  cursor: not-allowed;
}

.c10 .fi-button_loading-icon {
  display: inline-block;
  width: 20px;
  height: 20px;
  margin-right: 8px;
  animation: rotation 1.5s infinite linear;
}

.c10 .fi-button_loading-icon .fi-icon-component-brand-fill {
  fill: hsl(0, 0%, 100%);
}

.c10.fi-button--loading {
  display: flex;
  align-items: center;
}

.c10.fi-button--loading-icon-only .fi-button_loading-icon {
  margin-right: 0;
}

@media (prefers-reduced-motion) {
  .c10.fi-loadingSpinner.fi-loadingSpinner--loading .fi-loadingSpinner_icon {
    animation: rotation 10s infinite linear;
  }
}

<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    id="root"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="c0 fi-modal_base"
  >
    <div
      class="ReactModal__Overlay fi-modal_overlay"
    >
      <div
        aria-labelledby="test12"
        aria-modal="false"
        class="ReactModal__Content fi-modal"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="c1 fi-modal_content c2"
        >
          <span
            class="c3 c4 fi-modal_title"
          >
            <h2
              class="c5 fi-heading c6 fi-modal_title_focus-wrapper fi-heading--h3"
              tabindex="0"
            >
              Test modal
            </h2>
          </span>
          <p>
            Modal Content
          </p>
        </div>
        <div
          class="c1 c7 fi-modal_footer"
        >
          <div
            class="c1 fi-modal_footer_content"
          >
            <span
              aria-live="polite"
              class="c3 c8 fi-visually-hidden"
            />
            <button
              aria-disabled="false"
              class="c9 fi-button c10"
              tabindex="0"
              type="button"
            >
              OK
            </button>
          </div>
          <div
            class="c1 fi-modal_footer_content-gradient-overlay"
          >
            <div
              class="c1 fi-modal_footer_content-gradient"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
